planner/core/logical_plans.go

```diff
@@ -179,6 +181,77 @@ func (p *LogicalJoin) Shallow() *LogicalJoin {
 	return join.Init(p.ctx, p.blockOffset)
 }
 
+// ExtractFD implements the interface LogicalPlan.
+func (p *LogicalJoin) ExtractFD() *fd.FDSet {
+	switch p.JoinType {
+	case InnerJoin:
+		return p.extractFDForInnerJoin(nil)
+	case SemiJoin:
+		return p.extractFDForSemiJoin(nil)
+	default:
+		return &fd.FDSet{HashCodeToUniqueID: make(map[string]int)}
+	}
+}
+
+func (p *LogicalJoin) extractFDForSemiJoin(filtersFromApply []expression.Expression) *fd.FDSet {
+	// 1: since semi join will keep the part or all rows of the outer table, it's outer FD can be saved.
+	// 2: the un-projected column will be left for the upper layer projection or already be pruned from bottom up.
+	outerFD, _ := p.children[0].ExtractFD(), p.children[1].ExtractFD()
+	fds := outerFD
+
+	eqCondSlice := expression.ScalarFuncs2Exprs(p.EqualConditions)
+	allConds := append(eqCondSlice, p.OtherConditions...)
+	allConds = append(allConds, filtersFromApply...)
+	notNullColsFromFilters := extractNotNullFromConds(allConds, p)
+
+	constUniqueIDs := extractConstantCols(p.LeftConditions, p.SCtx(), fds)
+
+	fds.MakeNotNull(notNullColsFromFilters)
+	fds.AddConstants(constUniqueIDs)
+	p.fdSet = fds
+	return fds
+}
+
+func (p *LogicalJoin) extractFDForInnerJoin(filtersFromApply []expression.Expression) *fd.FDSet {
+	leftFD, rightFD := p.children[0].ExtractFD(), p.children[1].ExtractFD()
+	fds := leftFD
+	fds.MakeCartesianProduct(rightFD)
+
+	eqCondSlice := expression.ScalarFuncs2Exprs(p.EqualConditions)
+	// some join eq conditions are stored in the OtherConditions.
+	allConds := append(eqCondSlice, p.OtherConditions...)
+	allConds = append(allConds, filtersFromApply...)
+	notNullColsFromFilters := extractNotNullFromConds(allConds, p)
+
+	constUniqueIDs := extractConstantCols(allConds, p.SCtx(), fds)
+
+	equivUniqueIDs := extractEquivalenceCols(allConds, p.SCtx(), fds)
+
+	fds.MakeNotNull(notNullColsFromFilters)
+	fds.AddConstants(constUniqueIDs)
+	for _, equiv := range equivUniqueIDs {
+		fds.AddEquivalence(equiv[0], equiv[1])
+	}
+	// merge the not-null-cols/registered-map from both side together.
+	fds.NotNullCols.UnionWith(rightFD.NotNullCols)
+	if fds.HashCodeToUniqueID == nil {
+		fds.HashCodeToUniqueID = rightFD.HashCodeToUniqueID
+	} else {
+		for k, v := range rightFD.HashCodeToUniqueID {
+			if _, ok := fds.HashCodeToUniqueID[k]; ok {
+				panic("shouldn't be here, children has same expr while registered not only once")
+			}
+			fds.HashCodeToUniqueID[k] = v
+		}
+	}
+	for i, ok := rightFD.GroupByCols.Next(0); ok; i, ok = rightFD.GroupByCols.Next(i + 1) {
+		fds.GroupByCols.Insert(i)
+	}
+	fds.HasAggBuilt = fds.HasAggBuilt || rightFD.HasAggBuilt
```

Comment:
I think they should be put into `MakeCartesianProduct`.

Reply:
`MakeCartesianProduct ` will be also used in outer join. So we cannot put it into it.